code/game/objects/items/motion_detector.dm

```diff
@@ -1,179 +1,127 @@
+	. = ..()
+	for(var/datum/action/action AS in master_gun?.actions)
+		action.update_button_icon()
 
-	if(!detector_mode)
-		long_range_cooldown--
-		if(long_range_cooldown)
-			return
-		else
-			long_range_cooldown = initial(long_range_cooldown)
+/obj/item/attachable/motiondetector/update_icon_state()
+	icon_state = initial(icon_state) + (isnull(operator) ? "" : "_on")
 
-	if(ping)
-		playsound(loc, 'sound/items/detector.ogg', 60, 0, 7, 2)
+/// Signal handler to clean out user vars
+/obj/item/attachable/motiondetector/proc/clean_user()
+	SIGNAL_HANDLER
+	STOP_PROCESSING(SSobj, src)
+	if(operator)
+		UnregisterSignal(operator, list(COMSIG_PARENT_QDELETING, COMSIG_GUN_USER_UNSET))
+		UnregisterSignal(src, list(COMSIG_ITEM_EQUIPPED_TO_SLOT, COMSIG_ITEM_REMOVED_INVENTORY))
+		operator = null
+	QDEL_LIST_ASSOC_VAL(blip_pool)
+	update_icon()
 
-	var/detected
-	var/status
-	for(var/mob/living/M in orange(detector_range, operator))
-		if(!isturf(M.loc))
+/obj/item/attachable/motiondetector/process()
+	if(!operator?.client || operator.stat != CONSCIOUS)
+		clean_user()
+		return
+	closest_hostile_distance = MOTION_DETECTOR_RANGE + 1
+	for (var/mob/living/carbon/human/nearby_human AS in cheap_get_humans_near(operator, MOTION_DETECTOR_RANGE))
+		if(nearby_human == operator)
 			continue
-		status = MOTION_DETECTOR_HOSTILE //Reset the status to default
-		if(ishuman(M))
-			var/mob/living/carbon/human/H = M
-			if(!(H.wear_id?.iff_signal & iff_signal)) //device checks for IFF data and status
-				if(M.stat == DEAD)
-					if(H.has_working_organs() && H.get_ghost())
-						if(detect_revivable)
-							status = MOTION_DETECTOR_DEAD //Dead, but revivable.
-						else
-							continue
-					else
-						if(detect_fubar)
-							status = MOTION_DETECTOR_FUBAR //Dead and unrevivable; FUBAR
-						else
-							continue
-				else
-					if(detect_friendlies)
-						status = MOTION_DETECTOR_FRIENDLY
-					else
-						continue
-		if(world.time > M.last_move_time + 20 && (status == MOTION_DETECTOR_HOSTILE))
-			continue //hasn't moved recently
-
-		detected = TRUE
-
-		show_blip(operator, M, status)
-
-		if(detected && ping)
-			playsound(loc, 'sound/items/tick.ogg', 50, 0, 7, 2)
-
-
-/obj/item/motiondetector/proc/show_blip(mob/user, mob/target, status)
-	if(!user.client)
+		prepare_blip(nearby_human, nearby_human.stat != DEAD ? nearby_human.wear_id?.iff_signal & operator.wear_id.iff_signal ? MOTION_DETECTOR_FRIENDLY : MOTION_DETECTOR_HOSTILE : MOTION_DETECTOR_DEAD)
+	for (var/mob/living/carbon/xenomorph/nearby_xeno AS in cheap_get_xenos_near(operator, MOTION_DETECTOR_RANGE))
+		prepare_blip(nearby_xeno, nearby_xeno.stat == DEAD ? MOTION_DETECTOR_DEAD : MOTION_DETECTOR_HOSTILE)
+	if(closest_hostile_distance <= MOTION_DETECTOR_RANGE)
+		operator.playsound_local(loc, 'sound/items/tick.ogg', 100 - closest_hostile_distance * 3, 0, 7, 2)
+	recycle_countdown--
+	if(recycle_countdown <= 0)
+		recycle_countdown = MOTION_DETECTOR_RECYCLE_TIME
+		QDEL_LIST_ASSOC_VAL(blip_pool)
```

Comment:
Explain this

Reply:
To prevent from creating blips over and over, we store them and just move them around when needed.

We are never deleting them, even if the detected mob went too far.

Hence, every three minutes, we just clean everything to prevent having a list too big